src/psyclone/generator.py

```diff
@@ -254,14 +251,42 @@ def generate(filename, api="", kernel_paths=None, script_name=None,
     alg_gen = None
 
     if api == "gocean1.0":
-        # Create language-level PSyIR from the Fortran file
+        # Create language-level PSyIR from the Algorithm file
         reader = FortranReader()
         psyir = reader.psyir_from_file(filename)
 
         # Raise to Algorithm PSyIR
         alg_trans = AlgTrans()
         alg_trans.apply(psyir)
 
+        # For each kernel called from the algorithm layer
+        kernels = {}
+        for invoke in psyir.walk(AlgorithmInvokeCall):
+            kernels[id(invoke)] = {}
+            for kern in invoke.walk(KernelFunctor):
+
+                # Find the container that includes this kernel symbol
+                container_symbol = kern.symbol.interface.container_symbol
+
+                # Find the kernel file containing the container
+                filepath = get_kernel_filepath(
+                    container_symbol.name, kernel_paths, filename)
+
+                try:
+                    # Create language-level PSyIR from the kernel file
+                    kernel_psyir = reader.psyir_from_file(filepath)
+                except InternalError as info:
+                    print(f"In kernel file '{filepath}':\n{str(info.value)}",
+                          file=sys.stderr)
+                    sys.exit(1)
+
+                # Raise to Kernel PSyIR
+                kern_trans = KernTrans()
+                kern_trans.metadata_name = kern.symbol.name
```

Comment:
Ah, I understand now. Since you're creating a new KernTrans() for every kernel anyway, maybe this name should be an argument to the constructor instead? Alternatively, could it be passed as an argument to the `apply()` or is that then inconsistent with our general Transformation API?

Reply:
Passing via apply is inconsistent as you say. I could require it for the constructor but I thought it better to allow the same transformation to be able to be applied to different metadata, rather than tying it to particular metadata, even if that is not how it is currently used.